src/consensus/blockstore/slot_block_data.rs

```diff
@@ -155,94 +171,83 @@ impl BlockData {
         }
     }
 
-    /// Adds a new valid shred to the block.
-    ///
-    /// Assumes that the shred is valid, performs no more validity checks.
-    ///
-    /// Returns:
-    /// - `Some(VotorEvent::FirstShred)` if this was the first shred of this block,
-    /// - `Some(VotorEvent::Block)` if the block was successfully reconstructed,
-    /// - `None` otherwise.
-    pub fn add_valid_shred(&mut self, shred: Shred) -> Option<VotorEvent> {
+    fn add_shred(
+        &mut self,
+        shred: Shred,
+        leader_pk: PublicKey,
+    ) -> Result<Option<VotorEvent>, AddShredError> {
         assert!(shred.payload().header.slot == self.slot);
         let slice_index = shred.payload().header.slice_index;
-        let is_last_slice = shred.payload().header.is_last;
-        let is_first_shred = self.shreds.is_empty();
-        self.shreds.entry(slice_index).or_default().push(shred);
+        let cached_merkle_root = self.merkle_root_cache.entry(slice_index);
+        let validated_shred = ValidatedShred::new(shred, cached_merkle_root, &leader_pk)?;
+        self.add_validated_shred(validated_shred)
+    }
 
-        // store last slice index, delete everything after last slice
-        if is_last_slice && self.last_slice.is_none() {
-            self.last_slice = Some(slice_index);
-            self.slices.retain(|&ind, _| ind <= slice_index);
-            self.shreds.retain(|&ind, _| ind <= slice_index);
+    fn add_validated_shred(
+        &mut self,
+        validated_shred: ValidatedShred,
+    ) -> Result<Option<VotorEvent>, AddShredError> {
+        let shred = validated_shred.to_shred();
+        let header = &shred.payload().header;
+        assert!(header.slot == self.slot);
+        let slice_index = header.slice_index;
+
+        match (header.is_last, self.last_slice) {
+            (true, None) => {
+                self.last_slice = Some(slice_index);
+                self.slices.retain(|&ind, _| ind <= slice_index);
+                self.shreds.retain(|&ind, _| ind <= slice_index);
+            }
+            (true, Some(l)) => {
+                if slice_index != l {
+                    return Err(AddShredError::InvalidSliceIndex);
+                }
+            }
+            (false, None) => (),
+            (false, Some(l)) => {
+                if slice_index >= l {
+                    return Err(AddShredError::InvalidSliceIndex);
+                }
+            }
         }
 
-        // maybe send first shred notification
+        let is_first_shred = self.shreds.is_empty();
+        let slice_shreds = {
+            let shred_index = shred.payload().index_in_slice;
+            let slice_shreds = self.shreds.entry(slice_index).or_default();
+            let exists = slice_shreds
+                .iter()
+                .any(|s| s.payload().index_in_slice == shred_index);
```

Comment:
this always kind of bothers me, maybe it should rather be that `slice_shreds` holds something like `[Option<Shred>; TOTAL_SHREDS]`, maybe as a new PR...

Reply:
https://github.com/qkniep/alpenglow/issues/182